### ISSUE TITLE
Circular dependency detected for table with no foreign keys

### ISSUE THREAD

**Issue by TimSirmovics** [dotnet/efcore#33154]
The following code generates a Circular Dependency error for a single table with no foreign keys.
The structure of the table is from a legacy database where the Identity column is not the primary key.
The issue seems to occur when loading more than 1 entity, deleting those entities and then adding new entities with the same primary keys.

### Sample project and table setup:
[CircRefTest.zip](https://github.com/dotnet/efcore/files/14398055/CircRefTest.zip)
[TableSetup.sql.txt](https://github.com/dotnet/efcore/files/14398057/TableSetup.sql.txt)

### Stack trace

```
System.InvalidOperationException
  HResult=0x80131509
  Message=Unable to save changes because a circular dependency was detected in the data to be saved: 'CircRef { 'PkCol': 1 } [Added] <-
Index { 'IdCol': 0 } CircRef { 'PkCol': 2 } [Added] <-
Index { 'IdCol': 0 } CircRef { 'PkCol': 1 } [Added]'.
  Source=Microsoft.EntityFrameworkCore.Relational
  StackTrace:
   at Microsoft.EntityFrameworkCore.Utilities.Multigraph`2.ThrowCycle(List`1 cycle, Func`2 formatCycle, Func`2 formatException)
   at Microsoft.EntityFrameworkCore.Utilities.Multigraph`2.TopologicalSortCore(Boolean withBatching, Func`4 canBreakEdges, Func`2 formatCycle, Func`2 formatException)
   at Microsoft.EntityFrameworkCore.Utilities.Multigraph`2.BatchingTopologicalSort(Func`4 canBreakEdges, Func`2 formatCycle, Func`2 formatException)
   at Microsoft.EntityFrameworkCore.Update.Internal.CommandBatchPreparer.TopologicalSort(IEnumerable`1 commands)
   at Microsoft.EntityFrameworkCore.Update.Internal.CommandBatchPreparer.<BatchCommands>d__8.MoveNext()
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.Execute(IEnumerable`1 commandBatches, IRelationalConnection connection)
   at Microsoft.EntityFrameworkCore.Storage.RelationalDatabase.SaveChanges(IList`1 entries)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChanges(IList`1 entriesToSave)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChanges(StateManager stateManager, Boolean acceptAllChangesOnSuccess)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.<>c.<SaveChanges>b__112_0(DbContext _, ValueTuple`2 t)
   at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChanges(Boolean acceptAllChangesOnSuccess)
   at Microsoft.EntityFrameworkCore.DbContext.SaveChanges(Boolean acceptAllChangesOnSuccess)
   at Microsoft.EntityFrameworkCore.DbContext.SaveChanges()
   at Program.<Main>$(String[] args) in Program.cs:line 17
```

### Provider and version information

EF Core version: 8.0.2
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: net8.0
Operating system: Windows 11
IDE: Visual Studio 2022 17.9.1

**Comment by ajcvickers** [comment]
Note for team triage: Not a regression; still repros on main.

Model:

```
Model: 
  EntityType: CircRef
    Properties: 
      PkCol (int) Required PK AfterSave:Throw
      IdCol (int) Required Index ValueGenerated.OnAdd
    Keys: 
      PkCol PK
    Indexes: 
      IdCol Unique
```

Change tracker:

```
CircRef (Shared) {PkCol: 1} Added
    PkCol: 1 PK
    IdCol: 0
CircRef (Shared) {PkCol: 1} Deleted
    PkCol: 1 PK
    IdCol: 1
CircRef (Shared) {PkCol: 2} Added
    PkCol: 2 PK
    IdCol: 0
CircRef (Shared) {PkCol: 2} Deleted
    PkCol: 2 PK
    IdCol: 2
```

**Comment by ajcvickers** [comment]
@TimSirmovics We looked at this in triage, and the behavior here is because the database has a unique index on `IdCol`, but both entities have the same value. If the indexed properties have their values set, for example, with:

```
var newRecord1 = new CircRef { PkCol = 1, IdCol = 1 };
context.CircRefs.Add(newRecord1);

var newRecord2 = new CircRef { PkCol = 2, IdCol = 2 };
context.CircRefs.Add(newRecord2);
```

Then EF is both able to order the inserts, and the database won't throw a constraint violation.

**Comment by TimSirmovics** [comment]
@ajcvickers your example works where the `IdCol` matches what was previously in the table, but if I swap the values or use anything other than the original values, I get an exception `SqlException: Cannot update identity column 'IdCol'.`.

```
var newRecord1 = new CircRef { PkCol = 1, IdCol = 2 };
var newRecord2 = new CircRef { PkCol = 2, IdCol = 1 };

var newRecord1 = new CircRef { PkCol = 1, IdCol = 5 };
var newRecord2 = new CircRef { PkCol = 2, IdCol = 6 };
```

**Comment by ajcvickers** [comment]
@TimSirmovics That's a SQL Server limitation. SQL Server does not allow Identity columns to be updated.

**Comment by ajcvickers** [comment]
Note see also: #30705

**Comment by ajcvickers** [comment]
@TimSirmovics That is what #30705 covers.

**Comment by TimSirmovics** [comment]
Yes I think we posted at the same time.
I've just had a read, and it's probably a lack in my understanding but the linked example seems to be talking about where 2 entity types share the same table, and not a simple example like this with a single entity type.

In any case I see it's on the backlog so I've voted on the issue and issue grouping.

In the meantime as a workaround I will manually keep track of the identity values and re-apply them in the case of deleting and inserting entities with the same primary key.
